code/controllers/subsystem/weather.dm

```diff
@@ -8,24 +8,66 @@ SUBSYSTEM_DEF(weather)
 	var/list/eligible_zlevels = list()
 	var/list/next_hit_by_zlevel = list() //Used by barometers to know when the next storm is coming
 
-/datum/controller/subsystem/weather/fire()
+/datum/controller/subsystem/weather/fire(resumed = FALSE)
 	// process active weather
-	for(var/V in processing)
-		var/datum/weather/our_event = V
-		if(our_event.aesthetic || our_event.stage != MAIN_STAGE)
+	for(var/datum/weather/weather_event in processing)
+		if(IS_WEATHER_AESTHETIC(weather_event.weather_flags) || weather_event.stage != MAIN_STAGE)
 			continue
-		for(var/mob/act_on as anything in GLOB.mob_living_list)
-			if(our_event.can_weather_act(act_on))
-				our_event.weather_act(act_on)
+
+		if(weather_event.currentpart == SSWEATHER_MOBS)
+			if(!resumed)
+				weather_event.current_mobs = GLOB.mob_living_list.Copy()
+			var/list/current_mobs_cache = weather_event.current_mobs // cache for performance
+			while(current_mobs_cache.len)
+				var/mob/living/target = current_mobs_cache[current_mobs_cache.len]
+				current_mobs_cache.len--
+				if(QDELETED(target))
+					continue
+				if(weather_event.can_weather_act_mob(target))
+					weather_event.weather_act_mob(target)
+				if(MC_TICK_CHECK)
+					return
+			resumed = FALSE
+			weather_event.next_subsystem_task()
+
+		if(weather_event.currentpart == SSWEATHER_TURFS)
+			if(!resumed)
+				weather_event.turf_iteration = floor(weather_event.weather_turfs_per_tick)
+				var/fraction_chance = fract(weather_event.weather_turfs_per_tick) * 100
+				weather_event.turf_iteration += prob(fraction_chance) ? 1 : 0
+			var/list/weather_turfs = weather_event.weather_turfs // cache for performance
+			while(weather_event.turf_iteration)
+				weather_event.turf_iteration--
+				var/turf/open/selected_turf = pick(weather_turfs)
+				weather_event.weather_act_turf(selected_turf)
+				if(MC_TICK_CHECK)
+					return
+			resumed = FALSE
+			weather_event.next_subsystem_task()
+
+		if(weather_event.currentpart == SSWEATHER_THUNDER)
+			if(!resumed)
+				weather_event.thunder_iteration = floor(weather_event.thunder_turfs_per_tick)
+				var/fraction_chance = fract(weather_event.thunder_turfs_per_tick) * 100
+				weather_event.thunder_iteration += prob(fraction_chance) ? 1 : 0
```

Comment:
similar

Reply:
Same reasoning as above. If you want me to I can make this a macro somewhere since I think it's a useful but niche operation.